pam/gdm/protocol.go

```diff
@@ -0,0 +1,170 @@
+// Package gdm is the package for the GDM pam module handing.
+package gdm
+
+//TODO: Watch https://github.com/protocolbuffers/protobuf for any changes on the experimental status of optional fields,
+// previously described on: https://github.com/protocolbuffers/protobuf/blob/main/docs/implementing_proto3_presence.md.
+//
+// Should it become default, remove the --experimental_allow_proto3_optional flag from the go generate command below.
+//go:generate sh -c "PATH=\"$PATH:`go env GOPATH`/bin\" protoc --proto_path=. --go_out=. --go_opt=paths=source_relative gdm.proto --experimental_allow_proto3_optional -I../.."
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"slices"
+
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+const (
+	// ProtoVersion is the version of the JSON protocol.
+	ProtoVersion = uint32(1)
+)
+
+// Request is an interface implementing all the gdm requests.
+type Request = isRequestData_Data
+
+// Response is an interface implementing all the gdm responses.
+type Response = isResponseData_Data
+
+// Event is an interface implementing all the gdm events.
+type Event = isEventData_Data
+
+// NewDataFromJSON unmarshals data from json bytes.
+func NewDataFromJSON(bytes []byte) (*Data, error) {
+	var gdmData Data
+	if err := protojson.Unmarshal(bytes, &gdmData); err != nil {
+		return nil, err
+	}
+
+	if err := gdmData.Check(); err != nil {
+		return nil, err
+	}
+
+	return &gdmData, nil
+}
```

Comment:
Well this will be used by consumers inside the package, so technically it could just be not-exported, but I'm also planning to use it outside the `gdm` package as it's something I've been using so far for simulating the protocol in some integration tests.